### PR TITLE
Offer typed parameterized query helpers

### PR DESCRIPTION
**Pull request by yihezkel**
#### Pull Request Description

Offer typed parameterized query helpers

---

#### Future Release Comment
**Features:**
- Offer typed parameterized query helpers

### CODE REVIEW

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (yihezkel) [inline data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:3]: I considered using generics here so we can have 1 `T value` property, but didn't since we need specialized behavior for the subclasses.

**Comment by yihezkel** [comment]
I spoke with @ArielYehezkely and Amit Ofir, and we agreed that this use case is precisely something an SDK should offer - provide the user helper methods to make complex service formatting requirements transparent.

But I want to discuss this change with @AsafMah in terms of SDK parity.

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java
Comment (yogilad) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java:31]: TIMESPAN

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslBoolFormat.java
Comment (yogilad) [inline data/src/main/java/com/microsoft/azure/kusto/data/format/CslBoolFormat.java:26]: Why not simply `return value ? "true" : "false";`

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslDateTimeFormat.java
Comment (yogilad) [inline data/src/main/java/com/microsoft/azure/kusto/data/format/CslDateTimeFormat.java:13]: IIRC kusto can process partial datetime strings, e.g. this is valid query
`KustoLogs
| where Timestamp  between (datetime("2021") .. 20m)`

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslUuidFormat.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslUuidFormat.java:7]: On the one hand, it seems Java has no defined convention for this scenario.
https://stackoverflow.com/questions/2236807/java-naming-convention-with-acronyms brings up a few ideas, including:
- "For 4+ letter acronyms, use mixed case. The standard library does this, and it just makes good sense."
- Several "canonical" Java classnames, showing there's clearly no consistent approach.

On the other hand we import (literally 2 lines above) a class with the acronym-in-all-caps convention :-)
I would argue it makes sense there because it's **only** the acronym.

WDYT?

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java:82]: I'm leaning against it - I see this like setting a value in a `HashMap`, where the user is trusted to manage writing each key.

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java:31]: I'm naming these for the Kusto datatypes they're relevant to, and this is for the datatype called, "time".

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslBoolFormat.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslBoolFormat.java:26]: Because `Boolean` can be null, in which case testing it will throw a NPE.

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java:31]: Oops, I just checked out https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/scalar-data-types/ and see that the datatype's name is indeed "timespan". Weird, because I've only ever seen it as "time(...)".

I'll update the terminology accordingly.

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslDateTimeFormat.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslDateTimeFormat.java:13]: You're right, the Csl parser in C# is much more mature and can handle things like that. I'd suggest that achieving full feature parity with C# is out of scope in this PR, but I tried to reasonably mimic the C# DateTime behavior.

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (yihezkel) [inline data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:34]: I know this is unpopular in the "elegant" circles, but I personally consider this to be more readable.

Comment (AsafMah) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:34]: Do you mean the "no new line before else" part? 
Because that's fine in java

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: I'm pretty sure the ".*" is not needed

Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: We need it in case there's a whitespace character in the matching group.

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:34]: I think the popular eleagent approach is:
`if (matcher.matches()) {
return matcher.group(1);
}
return valueWithType;`

Comment (AsafMah) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:34]: oh, then I agree with the elegant approach. (with new lines)

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (AsafMah) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: I tried it in regex101, it's the same results with or without it.
Do you want to capture the whitespace inside the parens or not?

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:34]: I'm not a fan of its readability.

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: I haven't tried a tester, but I indeed want to capture any whitespace that's part of the value inside the parens.

Comment (AsafMah) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: That's part of the value = includes leading and trailing?
Those are your options (unescaped regexes):

```regex
int\s*\((.*)\)\s* 

int(aaaa) => "aaaa"
int( aaaa  )  => " aaaa  "
int(a a) => "a a"
int( a a ) => " a a "

int\s*\(\s*(.*?)\s*\)\s*

int(aaaa) => "aaaa"
int( aaaa  )  => "aaaa"
int(a a) => "a a"
int( a a ) => "a a"

```

Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: It doesn't make much sense for `int`, but consider others, like `datetime(2022-01-05 9:59:32)`

Comment (AsafMah) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: Ok I see what you mean, but I still prefer `\s*\(\s*(.*?)\s*\)\s*` (it's equivalent)

Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: Are you sure they're equivalent? I think mine enforces that there's some non-whitespace inside the capture group, and I don't think yours does.

Comment (AsafMah) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java:29]: Ok yes if that's something we want we can keep it as is

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java:15]: unstar

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java:31]: I feel like a regex this big deserves a comment with an explanation and example (or a link if you took it from somewhere)

Comment (AsafMah) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java:31]: Also I would prefer "PATTERN" over "REGEX"

File: data/src/main/java/com/microsoft/azure/kusto/data/KustoResultSetTable.java
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/KustoResultSetTable.java:301]: Why aren't the others also implemented this way?
```java
public int getInt(String columnName) {
        return getInt(findColumn(columnName));
        }
```

File: pom.xml
Comment (AsafMah) [inline pom.xml:43]: For what are we using this?

Comment (yihezkel) [reply pom.xml:43]: As discussed, for unescaping strings

File: data/src/main/java/com/microsoft/azure/kusto/data/KustoResultSetTable.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/KustoResultSetTable.java:301]: I looked briefly, and it looks like their call stacks are the same in practice, as `private Object get(String colName)` calls `public int findColumn(String columnName)`. It's just inconsistent how sometimes the `get[DataType]` method calls `findColumn` directly, and sometimes calls it via `get`.

**Review by AsafMah — APPROVED** [review]
Approved after offline discussion

File: data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java
Comment (yihezkel) [reply data/src/main/java/com/microsoft/azure/kusto/data/format/CslFormat.java]: Used ternary, as discussed

**Pull request merged**
